File: leetcode/p00110/p00110.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        
        def helper(node: Optional[TreeNode]) -> int:
            if not node:
                return 0
            
            ld = helper(node.left)
            rd = helper(node.right)
            if ld == -1 or rd == -1 or abs(ld-rd) > 1:
                return -1
            
            return max(ld, rd) + 1
        
        return helper(root) != -1

File: leetcode/p00110/p00110.go
package p00110

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time Complexity: O(n)
// Space Complexity: O(n)
func isBalanced(root *TreeNode) bool {
	return recursive(root) != -1
}

func recursive(node *TreeNode) float64 {
	if node == nil {
		return 0
	}

	ld := recursive(node.Left)
	rd := recursive(node.Right)
	if ld == -1 || rd == -1 || math.Abs(ld-rd) > 1 {
		return -1
	}

	return max(ld, rd) + 1
}

func max(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}
